TYP: `union1d` outputs `dtype[floating[_64Bit]]` which is incompatible with `dtype[float64]]`

### Describe the issue:

While type checking, `numpy.union1d` outputs an array of type `ndarray[tuple[int, ...], dtype[floating[_64Bit]]]` even if both the arguments are of the type `ndarray[tuple[int, ...], dtype[float64]]` -- failing mypy checks.

### Reproduce the code example:

```python
import numpy.typing as npt
import numpy as np


def do_something(a: npt.NDArray[np.float64], b: npt.NDArray[np.float64]) -> npt.NDArray[np.float64]:
    return np.union1d(a, b)
```

### Error message:

```shell
% mypy mypy_rep.py

mypy_rep.py:6: error: Incompatible return value type (got "ndarray[tuple[int, ...], dtype[floating[_64Bit]]]", expected "ndarray[tuple[int, ...], dtype[float64]]")  [return-value]
Found 1 error in 1 file (checked 1 source file)
```

### Python and NumPy Versions:

2.2.2
3.13.1 (main, Dec  3 2024, 17:59:52) [Clang 16.0.0 (clang-1600.0.26.4)]

---

mypy 1.14.1 (compiled: yes)
No extra configuration

### Runtime Environment:

[{'numpy_version': '2.2.2',
  'python': '3.13.1 (main, Dec  3 2024, 17:59:52) [Clang 16.0.0 '
            '(clang-1600.0.26.4)]',
  'uname': uname_result(system='Darwin', node='Saranshs-MacBook-Pro.local', release='24.2.0', version='Darwin Kernel Version 24.2.0: Fri Dec  6 19:02:41 PST 2024; root:xnu-11215.61.5~2/RELEASE_ARM64_T6030', machine='arm64')},
 {'simd_extensions': {'baseline': ['NEON', 'NEON_FP16', 'NEON_VFPV4', 'ASIMD'],
                      'found': ['ASIMDHP'],
                      'not_found': ['ASIMDFHM']}}]


### Context for the issue:

_No response_

--- a/numpy/lib/_arraysetops_impl.pyi
+++ b/numpy/lib/_arraysetops_impl.pyi
@@ -10,35 +10,7 @@ from typing import (
 from typing_extensions import deprecated
 
 import numpy as np
-from numpy import (
-    generic,
-    number,
-    ushort,
-    ubyte,
-    uintc,
-    uint,
-    ulonglong,
-    short,
-    int8,
-    byte,
-    intc,
-    int_,
-    intp,
-    longlong,
-    half,
-    single,
-    double,
-    longdouble,
-    csingle,
-    cdouble,
-    clongdouble,
-    timedelta64,
-    datetime64,
-    object_,
-    str_,
-    bytes_,
-    void,
-)
+from numpy import generic, number, int8, intp, timedelta64, object_
 
 from numpy._typing import (
     ArrayLike,
@@ -75,33 +47,17 @@ _NumberType = TypeVar("_NumberType", bound=number[Any])
 # Only relevant if two or more arguments are parametrized, (e.g. `setdiff1d`)
 # which could result in, for example, `int64` and `float64`producing a
 # `number[_64Bit]` array
-_SCTNoCast = TypeVar(
-    "_SCTNoCast",
+_EitherSCT = TypeVar(
+    "_EitherSCT",
     np.bool,
-    ushort,
-    ubyte,
-    uintc,
-    uint,
-    ulonglong,
-    short,
-    byte,
-    intc,
-    int_,
-    longlong,
-    half,
-    single,
-    double,
-    longdouble,
-    csingle,
-    cdouble,
-    clongdouble,
-    timedelta64,
-    datetime64,
-    object_,
-    str_,
-    bytes_,
-    void,
-)
+    np.int8, np.int16, np.int32, np.int64, np.intp,
+    np.uint8, np.uint16, np.uint32, np.uint64, np.uintp,
+    np.float16, np.float32, np.float64, np.longdouble,
+    np.complex64, np.complex128, np.clongdouble,
+    np.timedelta64, np.datetime64,
+    np.bytes_, np.str_, np.void, np.object_,
+    np.integer, np.floating, np.complexfloating, np.character,
+)  # fmt: skip
 
 class UniqueAllResult(NamedTuple, Generic[_SCT]):
     values: NDArray[_SCT]
@@ -339,11 +295,11 @@ def unique_values(x: ArrayLike, /) -> NDArray[Any]: ...
 
 @overload
 def intersect1d(
-    ar1: _ArrayLike[_SCTNoCast],
-    ar2: _ArrayLike[_SCTNoCast],
+    ar1: _ArrayLike[_EitherSCT],
+    ar2: _ArrayLike[_EitherSCT],
     assume_unique: bool = ...,
     return_indices: L[False] = ...,
-) -> NDArray[_SCTNoCast]: ...
+) -> NDArray[_EitherSCT]: ...
 @overload
 def intersect1d(
     ar1: ArrayLike,
@@ -353,11 +309,11 @@ def intersect1d(
 ) -> NDArray[Any]: ...
 @overload
 def intersect1d(
-    ar1: _ArrayLike[_SCTNoCast],
-    ar2: _ArrayLike[_SCTNoCast],
+    ar1: _ArrayLike[_EitherSCT],
+    ar2: _ArrayLike[_EitherSCT],
     assume_unique: bool = ...,
     return_indices: L[True] = ...,
-) -> tuple[NDArray[_SCTNoCast], NDArray[intp], NDArray[intp]]: ...
+) -> tuple[NDArray[_EitherSCT], NDArray[intp], NDArray[intp]]: ...
 @overload
 def intersect1d(
     ar1: ArrayLike,
@@ -368,10 +324,10 @@ def intersect1d(
 
 @overload
 def setxor1d(
-    ar1: _ArrayLike[_SCTNoCast],
-    ar2: _ArrayLike[_SCTNoCast],
+    ar1: _ArrayLike[_EitherSCT],
+    ar2: _ArrayLike[_EitherSCT],
     assume_unique: bool = ...,
-) -> NDArray[_SCTNoCast]: ...
+) -> NDArray[_EitherSCT]: ...
 @overload
 def setxor1d(
     ar1: ArrayLike,
@@ -400,9 +356,9 @@ def in1d(
 
 @overload
 def union1d(
-    ar1: _ArrayLike[_SCTNoCast],
-    ar2: _ArrayLike[_SCTNoCast],
-) -> NDArray[_SCTNoCast]: ...
+    ar1: _ArrayLike[_EitherSCT],
+    ar2: _ArrayLike[_EitherSCT],
+) -> NDArray[_EitherSCT]: ...
 @overload
 def union1d(
     ar1: ArrayLike,
@@ -411,10 +367,10 @@ def union1d(
 
 @overload
 def setdiff1d(
-    ar1: _ArrayLike[_SCTNoCast],
-    ar2: _ArrayLike[_SCTNoCast],
+    ar1: _ArrayLike[_EitherSCT],
+    ar2: _ArrayLike[_EitherSCT],
     assume_unique: bool = ...,
-) -> NDArray[_SCTNoCast]: ...
+) -> NDArray[_EitherSCT]: ...
 @overload
 def setdiff1d(
     ar1: ArrayLike,

--- a/numpy/typing/tests/data/reveal/arraysetops.pyi
+++ b/numpy/typing/tests/data/reveal/arraysetops.pyi
@@ -2,10 +2,7 @@ from typing import Any
 
 import numpy as np
 import numpy.typing as npt
-from numpy.lib._arraysetops_impl import (
-    UniqueAllResult, UniqueCountsResult, UniqueInverseResult
-)
-from numpy._typing import _64Bit
+from numpy.lib._arraysetops_impl import UniqueAllResult, UniqueCountsResult, UniqueInverseResult
 
 from typing_extensions import assert_type
 
@@ -28,7 +25,7 @@ assert_type(np.intersect1d(AR_M, AR_M, assume_unique=True), npt.NDArray[np.datet
 assert_type(np.intersect1d(AR_f8, AR_i8), npt.NDArray[Any])
 assert_type(
     np.intersect1d(AR_f8, AR_f8, return_indices=True),
-    tuple[npt.NDArray[np.floating[_64Bit]], npt.NDArray[np.intp], npt.NDArray[np.intp]],
+    tuple[npt.NDArray[np.float64], npt.NDArray[np.intp], npt.NDArray[np.intp]],
 )
 
 assert_type(np.setxor1d(AR_i8, AR_i8), npt.NDArray[np.int64])

